packages/private-apis/src/implementation.js

```diff
@@ -66,7 +66,7 @@ let allowReRegistration;
 // Let's default to true, then. Try/catch will fall back to "true" even if the
 // environment variable is not explicitly defined.
 try {
-	allowReRegistration = process.env.IS_WORDPRESS_CORE ? false : true;
+	allowReRegistration = globalThis.IS_WORDPRESS_CORE ? false : true;
```

Comment:
These `process.env`  variables are supposed to be tree shakable and if I'm not wrong web pack replaces `process.env.VARIABLE` with the right value in order to be able to tree-shake the build, which mean I'm not sure globalThis works for these variables.

Reply:
There isn't anything special about `process.env` that get special optimizations. Different compilers to provide some special defaults, but `process.env.WHATEVER` is likely more dangerous to use than `globalThis.WHATEVER`. Browsers don't have `process`, so using `process.env.SOMETHING` without an appropriately configured compiler has a strong chance of causing errors. https://github.com/WordPress/gutenberg/pull/16679 was a good example of this.

---

This PR is incomplete, but part of this involved fixing type issues with process. We claim some values are booleans, but this is [actually impossible in node](https://nodejs.org/docs/latest-v20.x/api/process.html#processenv). `process.env.ANYTHING` is type `undefined | string`:

>Assigning a property on process.env will implicitly convert the value to a string. This behavior is deprecated. Future versions of Node.js may throw an error when the value is not a string, number, or boolean.

It doesn't make much sense in the browser because the browser… doesn't have `process`. The thing that saves us is that this is handled by replacement in webpack, but it doesn't make a lot of sense to continue using webpack replacements in this way when there are better alternatives.

---

`globalThis` should be safer to use in published packages. Right now packages include some globals that are not very safe, like `SCRIPT_DEBUG`. Props to @gziolo for the `globalThis` idea.

In published packages, this isn't much different because no replacement is happening right now. In fact, this approach should be safer because `globalThis` _should be_ an object in all contexts I'm aware of and it will save needing to do checks like `typeof VAR !== 'undefined' && VAR` everywhere.

These things _do_ end up in packages published on npm, [for example the block-editor package](https://unpkg.com/browse/@wordpress/block-editor@12.25.0/build-module/components/block-controls/slot.js).

---

Regarding code elimination, I'll speak about webpack because it's what I'm most familiar with. I believe other systems work similarly.

_ASIDE_ I just noticed we have a [babel plugin](https://github.com/WordPress/gutenberg/blob/trunk/packages/warning/babel-plugin.js) that replaces the `warning()` package calls in the compiled packages so that they actually contain: `if ( typeof SCRIPT_DEBUG !== 'undefined' && SCRIPT_DEBUG === true ) {/*…*/}`. That also needs to be updated in this PR, so even where `SCRIPT_DEBUG` does not appear in source, it will appear in compiled packages for npm.

The systems must be configured to do replacement. Webpack does this with its `DefinePlugin`. That's why this PR includes these changed lines:

https://github.com/WordPress/gutenberg/blob/40c0fd7528d4d576d8e9e5ea96640f1eb4c2f4a1/tools/webpack/shared.js#L65-L74

As webpack compiles the code, it does replacement `DefinePlugin`. So when webpack encounters `globalThis.SCRIPT_DEBUG` it replaces it completely with `true`:

```js
if ( globalThis.SCRIPT_DEBUG ) { doThing() }
// becomes
if ( true ) { doThing() }
// or maybe
if ( false ) { doThing() }
```

The above example from the block editor package looks like this:
```js
// input
globalThis.SCRIPT_DEBUG === true ? warning(`Unknown BlockControls group "${group}" provided.`) : void 0;
// compiled (with webpack names replaced)
true ? warning(`Unknown BlockControls group "${group}" provided.`) : 0;
```

Depending on the compilation configuration, this may be the final result and what's included in the package. This is the first part of the process.

Some builds will optimize this code, which is where code is actually eliminated. After optimization, we wind up with:

```js
// true case
warning(`Unknown BlockControls group "${group}" provided.`);
// false case
// …condition and block are completely eliminated…
```

Reply:
I can confirm that tree-shaking works as expected with the changes applied.